### DISCUSSION TITLE
How to cascade delete properly in a self-relation with mongodb ?

### DISCUSSION THREAD

**Discussion by SimonChaumet** [prisma/prisma#23717] [Q&A]
### Question

It is written in the doc that we need to set NoAction on self-relation database but how are we supposed to do it ?

I tried getting all ids and using a deleteMany to delete parent and children together but I got a PrismaClientKnownRequestError.

### How to reproduce (optional)

1. Create self-relation
2. Try to delete a parent with their children

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model Tag {
  id           String   @id @default(auto()) @map("_id") @db.ObjectId
  name         String
  Parent       Tag?     @relation("Children", fields: [parentId], references: [id], onUpdate: NoAction, onDelete: NoAction)
  parentId     String?  @db.ObjectId
  Children     Tag[]    @relation("Children")
}
```

```ts
async function getAllChildren(
  prisma: PrismaClient,
  tag?: Prisma.TagGetPayload<object>
): Promise<Prisma.TagGetPayload<object>[]> {
  if (!tag) return [];
  const childrenTags = await prisma.tag.findMany({
    where: { parentId: tag.id },
  });
  const subTags = (
    await Promise.all(childrenTags.map((tag) => getAllChildren(prisma, tag)))
  ).flat();

  return [tag, ...subTags];
}

const allTags = await getAllChildren(
prisma,
await prisma.tag.findUnique({
  where: { id: request.params.tagId },
})
);

const tagsIds = allTags.map((t) => t.id);
return prisma.tag.deleteMany({
where: {
  id: { in: tagsIds },
},
});
```

- OS: ArchLinux
- Database: MongoDB in docker
- Node.js version: 20.11.1

Environment variables loaded from .env
prisma                  : 5.11.0
@prisma/client          : 5.11.0
Computed binaryTarget   : debian-openssl-1.0.x
Operating System        : linux
Architecture            : x64
Node.js                 : v20.11.1
Query Engine (Node-API) : libquery-engine efd2449663b3d73d637ea1fd226bafbcf45b3102 (at ../../node_modules/@prisma/engines/libquery_engine-debian-openssl-1.0.x.so.node)
Schema Engine           : schema-engine-cli efd2449663b3d73d637ea1fd226bafbcf45b3102 (at ../../node_modules/@prisma/engines/schema-engine-debian-openssl-1.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.11.0-15.efd2449663b3d73d637ea1fd226bafbcf45b3102
Default Engines Hash    : efd2449663b3d73d637ea1fd226bafbcf45b3102
Studio                  : 0.499.0

**Comment by ludralph** [comment]
Hi @SimonChaumet 👋 

Thank you for raising this question.

The `NoAction` referential action in Prisma Client is equivalent to `Restrict` in SQL, which means it prevents the deletion of a record if it's being referenced by another record. This is why you're getting a `PrismaClientKnownRequestError` when trying to delete a parent with their children.

If you want to delete both `parent` and `child` records, you would need to handle this in your application code as stated [here](https://github.com/prisma/prisma/issues/17649#issuecomment-1410518281). For example:
```js
const deleteTagAndChildren = async (parentId) => {
  const prisma = new PrismaClient()

  // Start a transaction
  const deletedRecords = await prisma.$transaction([
    prisma.tag.deleteMany({
      where: {
        parentId: parentId,
      },
    }),
    prisma.tag.delete({
      where: {
        id: parentId,
      },
    }),
  ])

  return deletedRecords
}

// Call the function with the ID of the parent you want to delete
deleteTagAndChildren('parent-id')
```
In the code above, notice how we delete the child first before deleting the parent.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by SimonChaumet** [reply]
Here I'm doing a recursive find to get all child ids, how does your code differs from mine except that mine is recursive ? Do I need to separate into multiple queries ?

**Comment by SimonChaumet** [comment]
Up please

**Comment by SimonChaumet** [comment]
Up again

**Comment by ludralph** [comment]
Hi @SimonChaumet 

After careful review, the standard `deleteMany` operation doesn't work for cascading deletes in self-relations. This is due to Prisma's built-in protections against violating referential integrity.  The most effective solution is to use MongoDB raw commands through Prisma's `$runCommandRaw` method. For example:
```js
const deleteTagAndChildren = async (parentId) => {
  const allChildren = await getAllChildren(parentId);
  const allIds = [parentId, ...allChildren.map(child => child.id)];

  try {
    const deletedRecords = await prisma.$runCommandRaw({
      delete: "Tag",
      deletes: [
        {
          q: { _id: { $in: allIds.map(id => ({ $oid: id })) } },
          limit: 0
        }
      ]
    });
    return deletedRecords;
  } catch (error) {
    console.error('Error deleting tags:', error);
    throw error;
  }
};

// Helper function to get all children recursively
async function getAllChildren(parentId) {
  const children = await prisma.tag.findMany({
    where: { parentId },
    select: { id: true },
  });

  const grandchildren = await Promise.all(
    children.map(child => getAllChildren(child.id))
  );

  return [...children, ...grandchildren.flat()];
}
```
This approach should work with MongoDB while achieving the cascading delete effect you're looking for. However, it's important to be aware that this is a workaround due to the current limitations of Prisma with MongoDB and self-relations.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
